### DISCUSSION TITLE
Search on multiple column with ilike

### DISCUSSION THREAD

**Discussion by ijaspreetsinghh** [supabase/supabase#6778] [Questions] [answered]
Hi, I just met with an issue where matching with ilike is only limited to one column, is there a way where we can match parameters of multiple columns on either primary or foreign table, it just needs to be optional for multiple columns.
Is there a way of matching multiple columns in ilike().

![image](https://user-images.githubusercontent.com/44422128/167852595-48574e5a-f94b-4f40-a03a-6ad6b7ae6e7f.png)

**Comment by GaryAustin1** [comment, marked as answer]
You can chain multiple ilikes (or any filter.  They will be anded. If you need an or then you use the or filter with multiple ilikes. 
https://supabase.com/docs/reference/dart/using-filters
https://supabase.com/docs/reference/javascript/using-filters#conditional-chaining

**Reply by ijaspreetsinghh** [reply]
Assuming 'OR' operation for '[Raw Postgrest syntax](https://postgrest.org/en/stable/api.html#horizontal-filtering-rows)', curl "http://localhost:3000/people?or=(age.lt.18,age.gt.21)"
->
or(column_1.ilike.%search%,column_2.ilike.%search%) OR or(column_1.ilike.\*search\*,column_2.ilike.\*search\*) this should work for me right?

**Reply by GaryAustin1** [reply]
I'm not sure about the special characters without trying it. But that is the right construct for an ilike or, with string after "ilike."

**Reply by ijaspreetsinghh** [reply]
![image](https://user-images.githubusercontent.com/44422128/168267335-15ecf1e2-7fcf-4f85-9a5c-ee9612ef24b0.png)
this worked for me, .or() ladder.

    or('or(name.ilike.%$search_term%,vendor_number.ilike.%$search_term%),or(email.ilike.%$search_term%,account.ilike.%$search_term%)')

**Reply by bennik88** [reply]
I'm new to this pattern and struggle a bit with returning the best match in the right order.
I want to search for `column` `first_name` and `column` `last_name` with one search input field where the user would enter combinations like `John`, `Doe`, `John Doe`, `Doe John`, `Jo`, `Do`. I've tried a lot of different combinations but when entering `John Doe` it always also lists results where `Doe` is not present (because of the `ilike.%search_tearm%). 
How would you approach this issue? 
Usually, I would think of writing an if clause, like if first_name and last_name given don't match only first_name or only last_name.
Also, is there a way to sort by best match?

**Herer is on of my code examples:**
```js
      if (searchText !== "") {
        const wordOne = searchText.trim().split(" ").at(0);
        const wordTwo = searchText.trim().split(" ").at(1);
        console.log({ wordOne, wordTwo });
        query = query.or(   `or(first_name.ilike.%${wordOne}%,last_name.ilike.%${wordTwo}%),and(first_name.ilike.%${wordTwo}%,last_name.ilike.%${wordOne}%)`
        );
      }
```

**Reply by GaryAustin1** [reply]
@bennik88 
You might want to ask a new question as there probably only two of us that will look at this older thread.
I think you could save yourself a lot of grief with just doing two inputs, a first name, last name ilike search and allow null for either (don't add the filter if they are null).  Otherwise you are probably looking at fulltext search or more with the commas and different formats.

**Reply by schachmatthias** [reply]
And here is the solution directly in the call, with pagination. If searchterm is empty all posts will be shown, so I didnt add a condition before, and it works just like a charm. Perhaps helpfull to some:
![image](https://github.com/supabase/supabase/assets/113443582/8eccff6b-24fb-4f33-9573-50fc588c6339)

Another possibility would define a Database function in Supabase:

![image](https://github.com/supabase/supabase/assets/113443582/38861d49-b1fb-4e0e-8003-68d7fc14d156)

You then can call this function in the frontend like so:
Beware that this is only fullTextSearch!

![image](https://github.com/supabase/supabase/assets/113443582/325feb26-83c6-4604-9a21-007f073aaedf)

This is in the pinia store of a Nuxt 3 application btw

**Comment by chrissy0** [comment]
Is there a way to filter for rows where ilike matches any element in an array column?

For example, the `my_numbers` column looks like this: `["one", "two", "three"]`

I want the row to be returned if any of the array elements contains `ree`.

**Reply by GaryAustin1** [reply]
Nothing built in.  The array operators only do full text compares.

One thing you could do if the array is not huge is use a computed field from PostgREST to make a comma separated string and do ilike on that.   No idea on performance for that.  

Otherwise you should search the web for Postgres partial match array column and see if there is something you can do with SQL.  I'm sure there is.

**Reply by chrissy0** [reply]
Thank you! The arrays usually have a length of 1-4. I went with option 1 and it worked like a charm. I had to recreate my view in order to "find" the new column.

**Comment by bkan36** [comment]
by the way if you want to apply the or on a related table here is the way:

```
            const { data, error } = await this.supabase
                .from('user_club')
                .select('user:user_id!inner(id, name, lastname)')
                .or(`name.ilike.%${searchQuery}%, lastname.ilike.%${searchQuery}%`, { referencedTable: 'user_id' })
```
user_club hold user_id and club_id

This example is used for a search bar

**Reply by Muneeb-Shah** [reply]
@bkan36 Thanks for sharing this. I've been finding this solution all day.

**Reply by Muneeb-Shah** [reply]
Just in case, if you have nested related tables, you can use . to access the nested relations. Here is an example:
```
   let query = supabase
      .from('calls')
      .select(
        `
        *,
        batch:batch_id!inner(
          name,
          property:property_id!inner(
            suburb
          )
        )
      `
      )
      .or(`name.ilike.%${searchQuery}%`, { referencedTable: 'batch_id)' });
      .or(`suburb.ilike.%${searchQuery}%`, { referencedTable: 'batch_id.property_id' // use . to access  inner relation 
      });
```

**Comment by lorikku** [comment]
What about sub-queries? I get `unexpected "t" expecting "not" or operator (eq, gt, ...)`

```
query.or(
      `title.ilike.%${title}%,transcripts.title.ilike.%${title}%`,
    );
```

is my query.

**Reply by GaryAustin1** [reply]
.or has to be all on one table.
There is a way to work around it for some cases: https://postgrest.org/en/latest/references/api/resource_embedding.html#or-filtering-across-embedded-resources

**Reply by Muneeb-Shah** [reply]
@bkan36  has already addressed this. Check his [comment](https://github.com/orgs/supabase/discussions/6778#discussioncomment-10270251)

**Reply by Muneeb-Shah** [reply]
@lorikku, update your query like this:

```
query
.or(`title.ilike.%${title}%`)
.or(`title.ilike.%${title}%`, { referencedTable: 'transcript_id', // rename it with your reference table field });
```

**Reply by GaryAustin1** [reply]
The solution bkan36 presents is for both or terms to be on the same referenced table.  The .or only does .or for columns inside the parameter passed to it.  All columns must be in that string separated by commas.
If you put two .or's one after the other they get anded together.  Because you only have one column in each .or your example is doing
.ilike(title,'%$(title}%') AND ilike (transcript_id.title,'%${title}%').

**Reply by Muneeb-Shah** [reply]
@GaryAustin1 Yes, you're right. It's Anding both or(). How can I make it work without Anding? Could you please suggest a solution?

**Reply by GaryAustin1** [reply]
There is no simple way to do that.
The link I provided above https://postgrest.org/en/latest/references/api/resource_embedding.html#or-filtering-across-embedded-resources shows a way that will work for some cases.  It is shown in the URL format (versus supabase-js) but it works in supabase-js with the correct filter not is null filter.

**Reply by Muneeb-Shah** [reply]
Thanks @GaryAustin1   

Based on the solution provided in that URL, how should this query be written:

query .or(`full_name.ilike.*test*`, { referencedTable: 'contact_id' }) .or(`suburb.ilike.*test*`, { referencedTable: 'batch_id.property_id' })

Could you please help me with that?

**Reply by GaryAustin1** [reply]
I don't know if it is even possible to go 2 levels deep with the linked method.
In other words I don't see how you can do batch_id.property_id.  
The or in the examples I linked only works on the main table columns to check that they are not null.  That won't work if one of those is another level deep.

You probably should use a view or rpc call for this and handle it in SQL.

**Reply by Muneeb-Shah** [reply]
Thanks @GaryAustin1, I will use view.

By the way 'batch_id.property_id' is actually working when using single or.

```javascript
let query = client
      .from('calls')
      .select(
        `
        *,
        batch:batch_id!inner(
          id,
          name,
          property:property_id!inner(
            id,
            suburb,
            street
          )
        )
      `,
        { count: 'exact' },
      )
      .eq('team.slug', accountSlug)
      .is('deleted_at', null);

      query = query
        .or(`suburb.ilike.*${term}*`, {
          referencedTable: 'batch_id.property_id',
        });
